x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Reverse.java

```diff
@@ -0,0 +1,167 @@
+                description = "`REVERSE` works with unicode, too! It keeps unicode grapheme clusters together during reversal."
+            ) }
+    )
+    public Reverse(
+        Source source,
+        @Param(
+            name = "str",
+            type = { "keyword", "text" },
+            description = "String expression. If `null`, the function returns `null`."
+        ) Expression field
+    ) {
+        super(source, field);
+        this.field = field;
+    }
+
+    private Reverse(StreamInput in) throws IOException {
+        this(Source.EMPTY, in.readNamedWriteable(Expression.class));
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(field());
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public DataType dataType() {
+        return field.dataType();
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        return isString(field, sourceText(), DEFAULT);
+    }
+
+    @Override
+    public boolean foldable() {
+        return field.foldable();
+    }
+
+    /**
+     * Reverses a unicode string, keeping grapheme clusters together
+     * @param str
+     * @return
+     */
+    public static String reverseStringWithUnicodeCharacters(String str) {
+        BreakIterator boundary = BreakIterator.getCharacterInstance(Locale.ROOT);
+        boundary.setText(str);
+
+        List<String> characters = new ArrayList<>();
+        int start = boundary.first();
+        for (int end = boundary.next(); end != BreakIterator.DONE; start = end, end = boundary.next()) {
+            characters.add(str.substring(start, end));
+        }
+
+        StringBuilder reversed = new StringBuilder(str.length());
+        for (int i = characters.size() - 1; i >= 0; i--) {
+            reversed.append(characters.get(i));
+        }
+
+        return reversed.toString();
+    }
+
+    private static boolean isOneByteUTF8(BytesRef ref) {
+        int end = ref.offset + ref.length;
+        for (int i = ref.offset; i < end; i++) {
+            if (ref.bytes[i] < 0) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static void reverseArray(byte[] array, int start, int end) {
+        while (start < end) {
+            byte temp = array[start];
+            array[start] = array[end];
+            array[end] = temp;
+            start++;
+            end--;
+        }
+    }
+
+    @Evaluator
+    static BytesRef process(BytesRef val) {
+        if (isOneByteUTF8(val)) {
+            // this is the fast path. we know we can just reverse the bytes.
+            BytesRef reversed = BytesRef.deepCopyOf(val);
+            reverseArray(reversed.bytes, reversed.offset, reversed.offset + reversed.length - 1);
```

Comment:
This scans an ASCII text for three times (even though not necessarily byte-by-byte).
I guess the first pass could attempt a reverse already, though that would imply allocation for non-ASCII text. 
I guess it's fine as is, probably not making a practical difference either way, can be optimised if ever needed.

Reply:
True. At least we're in O(n). 

> can be optimised if ever needed

Agreed